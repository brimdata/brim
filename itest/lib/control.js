"use strict";
exports.__esModule = true;
exports.retryUntil = exports.retry = void 0;
function retry(f, attempts, delay) {
    if (attempts === void 0) { attempts = 100; }
    if (delay === void 0) { delay = 100; }
    return new Promise(function (resolve, reject) {
        f().then(function (ret) { return resolve(ret); })["catch"](function (err) {
            setTimeout(function () {
                if (attempts === 1) {
                    reject(err);
                }
                else {
                    retry(f, attempts - 1, delay).then(function (ret) { return resolve(ret); })["catch"](function (err) { return reject(err); });
                }
            }, delay);
        });
    });
}
exports.retry = retry;
exports.retryUntil = function (f, cond_f, attempts, delay) {
    if (attempts === void 0) { attempts = 15; }
    if (delay === void 0) { delay = 1000; }
    // it is resolved and is responsible for examining and returning a boolean to
    // determine if the condition is satisfied.
    return retry(function () { return new Promise(function (resolve, reject) {
        f().then(function (val) {
            if (cond_f(val)) {
                resolve(val);
            }
            else {
                reject(new Error("retryUntil condition failure: " + val));
            }
        })["catch"](function (err) {
            reject("retryUntil promise failure: " + err);
        });
    }); }, attempts, delay);
};
