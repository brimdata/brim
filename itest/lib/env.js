/* @flow */

import path from "path"

export const isCI = (): boolean => process.env.GITHUB_ACTIONS === "true"

export const repoDir = (): string =>
  path.resolve(path.join(__dirname, "..", ".."))

export const itestDir = (): string =>
  path.join(process.env.WORKSPACE || "run", "itest")

export const testDataDir = (): string =>
  path.resolve(path.join(repoDir(), "itest", "testdata"))
