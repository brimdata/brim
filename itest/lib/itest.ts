import startApp from "./appStep/api/startApp"
import BrimDriver from "./brim-driver"
import BrimSelectors from "./brim-selectors"
import newAppInstance from "./newAppInstance"

export default function itest(name): [BrimDriver, BrimSelectors] {
  let app
  let testIdx = 0
  let brim = new BrimDriver(app)
  let $ = new BrimSelectors()

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    brim.app = app
    await startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return [brim, $]
}
