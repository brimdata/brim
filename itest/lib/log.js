"use strict"
exports.__esModule = true
exports.LOG = void 0
var path_1 = require("path")
var winston_1 = require("winston")
var env_1 = require("./env")
// When the flow-typed update produces an error....
// flow-typed/npm/winston_v3.x.x.js
// [1] 124â”‚     createLogger: <T>($winstonLoggerConfig<T>) => $winstonLogger<T>,
exports.LOG = winston_1.createLogger({
  level: "info",
  format: winston_1.format.combine(
    winston_1.format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss.SSS"
    }),
    winston_1.format.printf(function(_a) {
      var level = _a.level,
        message = _a.message,
        timestamp = _a.timestamp
      return "[" + timestamp + " " + level + "]: " + message
    })
  ),
  transports: [
    new winston_1.transports.Console({level: "info"}),
    // mkdir -p semantics so the environment variable seems OK enough.
    new winston_1.transports.File({
      filename: path_1["default"].join(env_1.itestDir(), "itest.log"),
      level: "debug"
    })
  ]
})
