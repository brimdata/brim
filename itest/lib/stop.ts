import fetch from "node-fetch"

export async function quitBrim(app) {
  if (app && app.isRunning()) {
    await app.stop()
    console.log("app stopped, waiting for zqd....")
    while (await isUp()) {
      await new Promise((go) => setTimeout(go, 500))
    }
    console.log("zqd is down")
  }
}

async function isUp() {
  try {
    const response = await fetch("http://localhost:9867/status")
    const text = await response.text()
    return text === "ok"
  } catch (e) {
    return false
  }
}
