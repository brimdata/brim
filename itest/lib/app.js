/* @flow */

import {retry} from "./control"
import {selectors} from "../../src/js/test/integration"
import {Application} from "spectron"

export const logIn = (app: Application) => {
  return app.client
    .setValue(selectors.login.host, "localhost")
    .setValue(selectors.login.port, "9867")
    .click(selectors.login.button)
}

export const waitForLoginAvailable = (app: Application) => {
  const waitForHostname = () => {
    return app.client.waitForExist(selectors.login.host)
  }
  const waitForPort = () => {
    return app.client.waitForExist(selectors.login.port)
  }
  const waitForButton = () => {
    return app.client.waitForExist(selectors.login.button)
  }
  return waitForHostname()
    .then(() => waitForPort())
    .then(() => waitForButton())
}

export const waitForSearch = (app: Application) => {
  return retry(() => app.client.element("#main-search-input").getValue())
}

export const waitForHistogram = (app: Application) => {
  return retry(() =>
    app.client.element(selectors.histogram.topLevel).getAttribute("class")
  )
}

export const writeSearch = (app: Application, searchText: string) =>
  app.client.setValue(selectors.search.input, searchText)

export const startSearch = (app: Application) =>
  app.client.click(selectors.search.button)

export const searchDisplay = async (app: Application) => {
  // This stinks. We have to use getHTML because headers that are off the
  // screen return as empty strings if you use getText. This isn't required of
  // actual results.
  // See http://v4.webdriver.io/api/property/getText.html
  // app.browserWindow.maximize() fixes the problem on my laptop but not CircleCI.
  // But what we get back includes the width which can be non-deterministic:
  // <div class="header-cell" style="width: 192px;">ts<div class="col-resizer"></div></div>
  // That style width will vary on my laptop vs. CircleCI.
  // The hack is to split this and extract just the text.
  const _trim = (s: string) => s.split(">")[1].split("<")[0]

  const headerResults = () => {
    return app.client.getHTML(selectors.viewer.headers).then((headers) => {
      if (typeof headers === "string") {
        headers = [headers]
      }
      return headers.map((h) => _trim(h))
    })
  }
  const searchResults = () => app.client.getText(selectors.viewer.results)

  let headers = await headerResults()
  let search = await searchResults()
  return headers.concat(search)
}

export const getCurrentSpace = (app: Application) =>
  app.client.getText(selectors.spaces.button)
