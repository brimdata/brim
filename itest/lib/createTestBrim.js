/* @flow */
import {
  click,
  ingestFile,
  newAppInstance,
  startApp,
  startSearch,
  writeSearch
} from "./app"
import {retryUntil} from "./control"
import lib from "../../src/js/lib"

export default (name: string) => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    await startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return {
    ingest(file: string) {
      return ingestFile(app, file)
    },

    mockSaveDialog(value: {canceled: boolean, filePath: ?string}) {
      return app.mainProcess.emit(
        "spectron:mock",
        "windows:showSaveDialog",
        value
      )
    },

    writeSearch(input: string) {
      return writeSearch(app, input)
    },

    startSearch() {
      return startSearch(app)
    },

    click(locator: string) {
      return click(app, locator)
    },

    waitForText(locator: string, regex: RegExp) {
      return retryUntil(
        () => app.client.getText(locator),
        (s) => regex.test(s)
      )
    },

    wait(ms: number) {
      return lib.sleep(ms)
    }
  }
}
