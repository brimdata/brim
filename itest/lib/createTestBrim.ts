import {Locator} from "../../src/js/test/createLocator"
import appStep from "./appStep/api"
// TODO in a future PR: remove direct logStep uses here.
import logStep from "./appStep/util/logStep"
import newAppInstance from "./newAppInstance"
import {retryUntil} from "./control"
import lib from "../../src/js/lib"

export default (name: string) => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    await appStep.startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return {
    getApp() {
      return app
    },

    ingest(file: string) {
      return appStep.ingestFile(app, file)
    },

    mockSaveDialog(value: {
      canceled: boolean
      filePath: string | null | undefined
    }) {
      return app.mainProcess.emit(
        "spectron:mock",
        "windows:showSaveDialog",
        value
      )
    },

    setValue(locator: Locator, value: string) {
      return logStep(`set input ${locator.css} to ${value}`, async () => {
        const el = await app.client.$(locator.css)
        await el.waitForDisplayed()
        return el.setValue(value)
      })
    },

    getText(locator: Locator) {
      return logStep(`get text from ${locator.css}`, async () => {
        const el = await app.client.$(locator.css)
        await el.waitForDisplayed()
        return el.getText()
      })
    },

    clickAppMenuItem(id: string) {
      return app.mainProcess.emit("spectron:clickAppMenuItem", id)
    },

    search(input: string) {
      return appStep.search(app, input)
    },

    click(locator: Locator) {
      return appStep.click(app, locator.css)
    },

    waitForText(locator: string, regex: RegExp) {
      return retryUntil(
        async () => (await app.client.$(locator)).getText(),
        (s) => regex.test(s)
      )
    },

    async isVisible(locator: Locator) {
      return !!(await (await app.client.$(locator.css)).isDisplayed())
    },

    async isNotVisible(locator: Locator) {
      return !(await (await app.client.$(locator.css)).isDisplayed())
    },

    wait(ms: number) {
      return lib.sleep(ms)
    }
  }
}
