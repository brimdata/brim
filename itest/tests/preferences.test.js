/* @flow */
import path from "path"

import {
  defaultModalButton,
  ingestProgressBar,
  jsonTypeConfigInput,
  viewerResults
} from "../../src/js/test/locators"
import {stdTest} from "../lib/jest"
import createTestBrim from "../lib/createTestBrim"
import testDataDir from "../lib/env"

const config = path.join(testDataDir(), "custom-schema.json")

describe("Preferences Modal Tests", () => {
  const brim = createTestBrim("preferences.test")

  stdTest("setting a custom json type config", async (done) => {
    brim.clickAppMenuItem("preferences")
    await brim.setValue(jsonTypeConfigInput, config)
    await brim.click(defaultModalButton)
    await brim.ingest("custom-sample.ndjson")
    await brim.waitUntil(() => brim.isNotVisible(ingestProgressBar))

    expect(await brim.getText(viewerResults)).toMatchSnapshot()
    done()
  })
})
