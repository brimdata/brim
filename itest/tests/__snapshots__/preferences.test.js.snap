// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preferences Modal Tests setting a custom json type config 1`] = `
"2020-02-25T16:03:17Z
capture_loss
11.854892015457153
zeek
0
6
0
2020-02-25T16:03:15Z
conn
CyE3UUwhn1u7RfHzl
192.168.1.110
57540
172.217.1.138
443
tcp
â¦»
0.06002688407897949
39
39
OTH
â¦»
â¦»
0
DdA
2
143
1
91
{
}
2020-02-25T16:03:15Z
conn
C96mYg3dHh6qk0tIKl
192.168.1.110
57572
172.217.6.138
443
tcp
â¦»
0.05391693115234375
63
0
SF
â¦»
â¦»
0
DFafA
4
271
3
156
{
}
2020-02-25T16:03:15Z
conn
CW0ZEh15E80oiZMuF4
192.168.1.179
51524
192.168.1.255
15600
udp
â¦»
â¦»
â¦»
â¦»
S0
â¦»
â¦»
0
D
1
63
0
0
{
}
2020-02-25T16:03:14Z
files
FMa5RD1f8J80UyJKQe
{
209.216.230.240
}
{
192.168.1.110
}
{
CHpIVG1QXAEWTym6v
}
SSL
0
{
MD5
,
SHA1
}
application/ocsp-response
â¦»
0
â¦»
F
471
â¦»
0
0
F
â¦»
fca4341c673e74dc3e330e1640e9b7d4
570304f7776ed6ca29464404554fb927395328e6
â¦»
â¦»
â¦»
â¦»
2020-02-25T16:03:14Z
files
FXkLzR25cwYgeI5yI7
{
209.216.230.240
}
{
192.168.1.110
}
{
CHpIVG1QXAEWTym6v
}
SSL
0
{
MD5
,
SHA1
,
X509
}
application/x-x509-user-cert
â¦»
0
â¦»
F
1,711
â¦»
0
0
F
â¦»
d16bf2ff6647cb8bccce2594e9237e50
987050ffb905cad3a79a8596c2120db97c03a165
â¦»
â¦»
â¦»
â¦»
2020-02-25T16:03:14Z
files
Frbo5s1CRsV02JUoSe
{
209.216.230.240
}
{
192.168.1.110
}
{
CHpIVG1QXAEWTym6v
}
SSL
0
{
MD5
,
SHA1
,
X509
}
application/x-x509-ca-cert
â¦»
0
â¦»
F
1,176
â¦»
0
0
F
â¦»
345eff15b7a49add451b65a7f4bdc6ae
1fb86b1168ec743154062e8c9cc5b171a4b7ccb4
â¦»
â¦»
â¦»
â¦»
2020-02-25T16:03:14Z
files
F1VUER3CattO8PiHWc
{
209.216.230.240
}
{
192.168.1.110
}
{
CHpIVG1QXAEWTym6v
}
SSL
0
{
MD5
,
SHA1
,
X509
}
application/x-x509-ca-cert
â¦»
0
â¦»
F
947
â¦»
0
0
F
â¦»
79e4a9840d7d3a96d7c04fe2434c892e
a8985d3a65e5e5c4b2d7d66d40c6dd2fb19c5436
â¦»
â¦»
â¦»
â¦»
2020-02-25T16:03:14Z
x509
FXkLzR25cwYgeI5yI7
3
074FE902C6B7D619884E1CA1854D9178
CN=news.ycombinator.com,O=Y Combinator\\\\\\\\, Inc.,L=San Francisco,ST=California,C=US
CN=DigiCert SHA2 Secure Server CA,O=DigiCert Inc,C=US
2019-07-08T07:00:00Z
2021-09-10T19:00:00Z
rsaEncryption
sha256WithRSAEncryption
rsa
2,048
65537
â¦»
[
news.ycombinator.com
]
[
]
[
]
[
]
F
â¦»
2020-02-25T16:03:14Z
ssl
CHpIVG1QXAEWTym6v
192.168.1.110
57640
209.216.230.240
443
TLSv12
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
secp256r1
news.ycombinator.com
F
â¦»
http/1.1
T
[
FXkLzR25cwYgeI5yI7
,
Frbo5s1CRsV02JUoSe
,
F1VUER3CattO8PiHWc
]
[
]
CN=news.ycombinator.com,O=Y Combinator\\\\\\\\, Inc.,L=San Francisco,ST=California,C=US
CN=DigiCert SHA2 Secure Server CA,O=DigiCert Inc,C=US
â¦»
â¦»
ok
2020-02-25T16:03:13Z
conn
CHpIVG1QXAEWTym6v
192.168.1.110
57640
209.216.230.240
443
tcp
ssl
0.10462594032287598
1,088
6,425
S1
â¦»
â¦»
0
ShADda
10
1,620
8
6,849
{
}
2020-02-25T16:03:13Z
conn
Cof4bD2JRa0kXKZo5k
192.168.1.110
51848
192.168.1.254
53
udp
dns
0.0169069766998291
38
54
SF
â¦»
â¦»
0
Dd
1
66
1
82
{
}
2020-02-25T16:03:13Z
dns
Cof4bD2JRa0kXKZo5k
192.168.1.110
51848
192.168.1.254
53
udp
47,856
0.0169069766998291
news.ycombinator.com
1
C_INTERNET
1
A
0
NOERROR
F
F
T
T
0
[
209.216.230.240
]
[
35
]
F
2020-02-25T16:03:12Z
conn
CR0J3f0wL63nIM5ga
192.168.1.110
57635
17.125.252.5
443
tcp
â¦»
0.5938777923583984
699
305
OTH
â¦»
â¦»
0
DadA
5
899
5
505
{
}
2020-02-25T16:03:11Z
conn
CQ137z1tDVuWxqRxR2
192.168.1.110
54375
192.168.1.254
53
udp
dns
0.01794600486755371
46
62
SF
â¦»
â¦»
0
Dd
1
74
1
90
{
}
2020-02-25T16:03:11Z
dns
CQ137z1tDVuWxqRxR2
192.168.1.110
54375
192.168.1.254
53
udp
28,084
0.01794600486755371
11.client-channel.google.com
1
C_INTERNET
1
A
0
NOERROR
F
F
T
T
0
[
173.194.201.189
]
[
213
]
F
2020-02-25T16:03:10Z
conn
C2vpb52uHsiH1lrhY5
192.168.1.110
55351
18.205.93.211
443
tcp
â¦»
0.08867907524108887
215
193
OTH
â¦»
â¦»
0
DadA
2
319
2
297
{
}
2020-02-25T16:03:10Z
conn
CjZSaP1lgDVf4GsIN3
192.168.1.110
57332
64.233.179.189
443
tcp
â¦»
1.0791869163513184
363
382
OTH
â¦»
â¦»
0
^dADa
7
727
7
746
{
}
2020-02-25T16:03:09Z
conn
CW7RMQqzAtSG0N7P1
192.168.1.179
47783
192.168.1.255
15600
udp
â¦»
â¦»
â¦»
â¦»
S0
â¦»
â¦»
0
D
1
63
0
0
{
}
2020-02-25T16:03:08Z
conn
CvdVQi4kuEGcaK6dq2
192.168.1.110
55354
52.37.243.173
443
tcp
â¦»
0.7618169784545898
114
56
OTH
â¦»
â¦»
0
DdAa
3
270
2
160
{
}
2020-02-25T16:03:08Z
conn
Ccr57r21XgKVazIICf
192.168.1.110
57487
192.30.253.125
443
tcp
â¦»
0.07794380187988281
28
24
OTH
â¦»
â¦»
0
^dADa
2
132
2
128
{
}
2020-02-25T16:03:08Z
conn
CiAOe94TJ2ei2zd6Cg
192.168.1.110
57326
173.194.201.189
443
tcp
â¦»
1.152367115020752
403
380
OTH
â¦»
â¦»
0
^dADa
7
767
7
744
{
}
2020-02-25T16:03:08Z
conn
Cq24042YRBPRCjYw73
192.168.1.110
55346
52.37.243.173
443
tcp
â¦»
1.2302970886230469
114
56
OTH
â¦»
â¦»
0
DdAa
3
270
2
160
{
}
2020-02-25T16:03:08Z
conn
CMDC9B1cyR6VNxwPag
192.168.1.110
55344
52.37.243.173
443
tcp
â¦»
0.03426098823547363
56
56
OTH
â¦»
â¦»
0
DadA
2
160
2
160
{
}
2020-02-25T16:03:07Z
conn
CvXyT7zWJ8Z8laJxg
192.168.1.110
55747
13.52.5.22
443
tcp
â¦»
0.017642974853515625
39
39
OTH
â¦»
â¦»
0
DdA
2
143
1
91
{
}
2020-02-25T16:03:07Z
conn
CWJsEz44vPePOSfllj
192.168.1.110
55635
18.246.31.137
443
tcp
â¦»
0.040701866149902344
39
39
OTH
â¦»
â¦»
0
DdA
2
143
1
91
{
}
2020-02-25T16:03:06Z
weird
â¦»
â¦»
â¦»
â¦»
â¦»
unknown_protocol
2
F
zeek
2020-02-25T16:03:06Z
conn
CaaI2s2TPbLyAmLsoh
192.168.1.110
57591
172.217.9.142
443
tcp
â¦»
0.21641206741333008
2,511
3,263
OTH
â¦»
â¦»
0
DadA
11
3,083
9
3,731
{
}
2020-02-25T16:03:05Z
conn
CFRnBd4kFJEPw0NIyd
192.168.1.110
56625
172.217.9.142
443
tcp
â¦»
11.317513942718506
2,801
1,306
OTH
â¦»
â¦»
0
DadA
16
3,633
16
2,138
{
}
2020-02-25T16:03:05Z
stats
zeek
72
1
1,001
â¦»
â¦»
â¦»
405
12
1
0
0
1
0
0
38
34
0
0
0
0
1,008
0
0
0
End of results ðŸŽ‰"
`;
